test/test_corpus.cc

```diff
@@ -601,6 +602,110 @@ TEST_P(LSPTest, All) {
+                    auto &response = msg->asResponse();
+                    ASSERT_TRUE(response.result) << "Code action request returned error: " << msg->toJSON();
+                    auto &receivedCodeActionResponse =
+                        get<variant<JSONNullObject, vector<unique_ptr<CodeAction>>>>(*response.result);
+
+                    EXPECT_FALSE(get_if<JSONNullObject>(&receivedCodeActionResponse));
+
+                    if (!get_if<JSONNullObject>(&receivedCodeActionResponse)) {
+                        UnorderedMap<string, unique_ptr<CodeAction>> receivedCodeActionsByTitle;
+                        auto &receivedCodeActions = get<vector<unique_ptr<CodeAction>>>(receivedCodeActionResponse);
+                        for (auto &codeAction : receivedCodeActions) {
+                            bool codeActionTitleUnique =
+                                receivedCodeActionsByTitle.find(codeAction->title) == receivedCodeActionsByTitle.end();
+                            EXPECT_TRUE(codeActionTitleUnique)
+                                << "Found code action with duplicate title: " << codeAction->title;
+
+                            if (codeActionTitleUnique) {
+                                receivedCodeActionsByTitle[codeAction->title] = move(codeAction);
+                            }
+                        }
+
+                        vector<shared_ptr<ApplyCodeActionAssertion>> matchedCodeActionAssertions;
+
+                        // Check any code action assertions for the same error range, i.e. verify that when applied,
+                        // they match the expectation.
+                        auto it = applyCodeActionAssertions.begin();
+                        while (it != applyCodeActionAssertions.end()) {
+                            auto codeActionAssertion = it->get();
+                            // TODO(sushain): what if there are multiple errors on the same range?
+                            if (cmpRanges(*codeActionAssertion->range, *error->range)) {
+                                auto it2 = receivedCodeActionsByTitle.find(codeActionAssertion->title);
+                                EXPECT_NE(it2, receivedCodeActionsByTitle.end());
+                                if (it2 != receivedCodeActionsByTitle.end()) {
+                                    auto codeAction = move(it2->second);
+                                    codeActionAssertion->check(test.sourceFileContents, codeAction, test.testName,
+                                                               fileUri);
+                                    matchedCodeActionAssertions.emplace_back(*it);
+                                    it = applyCodeActionAssertions.erase(it);
+                                    continue;
+                                }
+                            }
+
+                            ++it;
+                        }
+
+                        EXPECT_TRUE(!exhaustiveApplyCodeAction ||
```

Comment:
If I'm reading this correctly, this checks if we received all of the expected code actions for the given error.

Can we have separate conditions for the following with decent error messages:

* Received an unexpected code action.
* Did not receive expected code action.